### DISCUSSION TITLE
Do we need `R.trampoline` for tail-call elimination?

### DISCUSSION THREAD

**Discussion by semmel** [ramda/ramda#3456] [Ideas]
Recursive functions are integral to FP.

8 years after ES6, [Tail-Call Elimination][ragan] is just [supported by WebKit][eriklangille] — so on Safari and Bun JS runtimes.
On other runtimes, `trampoline` is needed to avoid stack overflow errors.
 
Since it would be a great use-case for `R.thunkify`, `trampoline` too would be a good fit for Ramda. IMO.

### History:
- `R.trampoline` was already suggested in #1484.
- `RA.trampoline` was attempted in [#1425][ra-1425]

### Usage:

```javascript
import { thunkify } from 'ramda';
// A tail-call recursive function, needs to return its thunk.
// fib :: (Number, Number, Number) → ( Number | () → ( Number | () → ( Number | … ) ) )
const fib = thunkify((n, k = 0, m = 1) => n === 1 ? m : fib(n - 1, m, k + m));

// trampoline as higher-order function
// fibT :: Number → Number
const fibT = trampoline(fib)
fibT(10); //=> 55

// trampoline as direct function application
trampolineApp(fib(10)); //=> 55
```

### Signature

As high-order function returning an auto-curried function the signature could be something like
`trampoline :: ((a, b, …, j) → Thunk k) → a → b → … → f → k`
`   Thunk k = (() → (k | () → (k | … )))`

Otherwise when simply applying the recursion 
`trampolineApp :: Thunk a → a`

I'd prefer the higher-order function, though.

The [implementation][jsitor] is straight forward.

Do we need this?

[eriklangille]: https://eriklangille.com/blog/proper_tail_calls.html
[ragan]: https://raganwald.com/2013/03/29/high-level-trampolining.html
[ra-1425]: https://github.com/char0n/ramda-adjunct/pull/1425
[jsitor]: https://jsitor.com/cJPBbe1nyNV

**Comment by CrossEye** [comment]
I think this would be a very good idea.

We should have done it years ago, but we did hope that the ES6 specifications would win out and TCO would be implemented.  Instead the vendors are claiming it's too problematic, even after Safari demonstrated that it's not a real issue.

I think we'd have to come up with a different implementation than the linked

```js
  trampoline = pc => {
    while (typeof pc === 'function') {
      pc = pc();
    }
    return pc;
  },
```

This treats functions as different from other values, with the implicit assumption that a recursive function couldn't return a function.

I'd have to think about how best to implement one, though.

**Comment by CrossEye** [comment]
Maybe related, a [gist of mine](https://gist.github.com/CrossEye/f28150a5f2f82bcc93106e262c0d0dcd) using a Loop-Recur implementation adapted from one seen on StackOverflow:

```js
const Loop = (fn, ... init) => { 
  let r = fn (... init)
  while (r instanceof Recur) r = fn (... r)
  return r
}

const Recur = ( ...v) => Object .create (
  Recur .prototype, 
  {[Symbol .iterator]: {value: _ => v .values ()}}
)
```

**Reply by semmel** [reply]
> (The implementation) treats functions as different from other values, with the implicit assumption that a recursive function couldn't return a function. 
>
> I'd have to think about how best to implement one, though.

Indeed that is not good! I had a slight feeling that it was too easy.

Perhaps we can do it analog to [`chainRec`][fl-chainRec]?
Like
```javascript
const fib = (next, done, n, k = 0, m = 1) => n === 1 ? done(m) : next(n - 1, m, k + m);
```

So, `trampoline(fn)(x)` would be something like `Identity.chainRec(fn, x)`? 
Perhaps the Ramda users are already accustomed to rewrite their recursive functions in that chainRec style `(next, done, x) => ` `trampoline :: ((a → c a b, b → c a b, a) → c a b, a) → b` 


[fl-chainRec]: https://github.com/fantasyland/fantasy-land/issues/250
